alternative-1-iterate-generic.Rmd

```diff
@@ -78,81 +82,109 @@ iterate.SquaresSequence <- function(x, state = NULL) {
     state <- 1L
   if (state > x$to)
     return(NULL)
-  list(value = (x$from - 1L + state) ^ 2, 
-       state = state + 1L)
+  list(
+    value = (x$from - 1L + state) ^ 2, 
+    state = state + 1L
+  )
 }
 
-for (x in SquaresSequence(1, 10))
+for (x in SquaresSequence(5, 10))
   print(x)
 ```
 
-Reticulate support might look like:
+## Sequence of unknown length
 
 ```{r}
-iterate.python.builtin.object <- function(x, state = NULL) {
-  if(is.null(state))
-    state <- reticulate::as_iterator(x)
+SampleSequence <- function(max) {
+  structure(
+    list(max = max),
+    class = "SampleSequence"
+  )
+}
 
-  sentinal <- environment()
-  value <- reticulate::iter_next(state, completed = sentinal)
-  if(identical(value, sentinal))
-    NULL
-  else
-    list(value, state)
+iterate.SampleSequence <- function(x, state = NULL) {
+  if(is.null(state))
+    state <- 0L
+  if (state > x$max)
+    return(NULL)
+  
+  val <- abs(rnorm(1))
+  list(
+    value = val, 
+    state = state + val
+  )
 }
 
-for(x in reticulate::r_to_py(1:3))
+for (x in SampleSequence(2)) {
   print(x)
+}
 ```
 
-The above is a narrow but flexible change (Only one new symbol would be
-introduced, `base::iterate()`). Since `iterate()` is a new generic, it
-would allows for changes in behavior to be made with intention
-individually for each class type.
+## Extensions
 
-## Optional extensions
+### Support for S3 classes
 
-### `iterate.function()`
+A simple modification to the default iterator (from `.subset2()` to
+`[[`) makes it possible to selectively change the behavior of key S3
+classes:
 
-An optional extension might be for the `base` namespace to also provide
-a `iterate()` method for functions. This would allow users to pass
-functions directly to `for`.
+```{r}
+iterate.POSIXt <- function(x, state = NULL) {
```

Comment:
I dropped the examples for factors and version numbers since they use the same code.